system/lib/compiler-rt/stack_ops64.s

```diff
@@ -0,0 +1,60 @@
+.globl stackSave
+.globl stackRestore
+.globl stackAlloc
+.globl emscripten_stack_init
+.globl emscripten_stack_get_current
+.globl emscripten_stack_get_free
+
+.globaltype __stack_pointer, i64
+
+stackSave:
+  .functype stackSave() -> (i64)
+  global.get __stack_pointer
+  end_function
+
+stackRestore:
+  .functype stackRestore(i64) -> ()
+  local.get 0
+  global.set __stack_pointer
+  end_function
+
+stackAlloc:
+  .functype stackAlloc(i64) -> (i64)
+  .local i64, i64
+  global.get __stack_pointer
+  # Get arg 0 -> number of bytes to allocate
+  local.get 0
+  # Stack grows down.  Subtract arg0 from __stack_pointer
+  i64.sub
+  # Align result by anding with ~15
+  i64.const 0xfffffff0
+  i64.and
+  local.tee 1
+  global.set __stack_pointer
+  local.get 1
+  end_function
+
+emscripten_stack_get_current:
+  .functype emscripten_stack_get_current () -> (i64)
+  global.get __stack_pointer
+  end_function
+
+.functype emscripten_stack_get_end () -> (i64)
+.functype emscripten_stack_get_base () -> (i64)
+.globaltype __stack_end, i64
+__stack_end:
+
+emscripten_stack_init:
+  # initialize __stack_end such that future calls to emscripten_stack_get_free
+  # use the correct value.
+  .functype emscripten_stack_init () -> ()
+  call emscripten_stack_get_end
+  global.set __stack_end
+  end_function
+
+emscripten_stack_get_free:
+  .functype emscripten_stack_get_free () -> (i64)
+  global.get __stack_pointer
+  global.get __stack_end
+  i64.sub
+  end_function
```

Comment:
I'm not totally against a new file but I think you could also make this `.S` (capital S means run the C pre-processor) file and use `ifdefs` and macros maybe?

Reply:
I remember we mentioned implementing these functions in a different way.. can you remind me how?

https://github.com/emscripten-core/emscripten/blob/c92bbd565a8c0732dbd1fc2dfccd78a3c58a89e3/system/lib/pthread/library_pthread_stub.c#L301-L320

Reply:
And how would the preprocessor know about the `MEMORY64` flag passed to `emcc` ?
See the change in `tools/system_libs.py` below for how this is currently included in the build system.

Reply:
`#ifdef __wasm64__` I think will work.